File: app/src/main/java/com/prey/events/manager/LocationLowBatteryRunner.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2015 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.events.manager;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import android.content.Context;

import com.prey.PreyConfig;
import com.prey.PreyLogger;

public class LocationLowBatteryRunner  {

    private static  SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yy hh:mm:ss", Locale.getDefault());

    public static boolean isValid(Context ctx) {
        try {
            Calendar cal = Calendar.getInstance();
            cal.setTime(new Date());
            cal.add(Calendar.HOUR, -3);
            long leastSixHours = cal.getTimeInMillis();
            long locationLowBatteryDate = PreyConfig.getPreyConfig(ctx).getLocationLowBatteryDate();
            PreyLogger.d("EVENT locationLowBatteryDate :" + locationLowBatteryDate + " " + sdf.format(new Date(locationLowBatteryDate)));
            PreyLogger.d("EVENT leastSixHours   :" + leastSixHours + " " + sdf.format(new Date(leastSixHours)));
            PreyLogger.d("EVENT diff:"+(leastSixHours-locationLowBatteryDate));
            if (locationLowBatteryDate == 0 || leastSixHours > locationLowBatteryDate) {
                long now = new Date().getTime();
                PreyConfig.getPreyConfig(ctx).setLocationLowBatteryDate(now);
                return true;
            }
            return false;
        } catch (Exception e) {
            return false;
        }
    }

}

File: app/src/main/java/com/prey/events/manager/LocationLowBatteryRunner.kt
package com.prey.events.manager

import android.content.Context
import com.prey.PreyConfig
import com.prey.PreyLogger
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Singleton object responsible for managing and throttling "Low Battery" location events.
 *
 * This runner ensures that location-based low battery events are not triggered too frequently,
 * implementing a cooldown period (3 hours) between consecutive valid events to optimize
 * battery usage and prevent redundant notifications.
 */
object LocationLowBatteryRunner {

    private const val THREE_HOURS_MS = 3 * 60 * 60 * 1000L
    private const val DATE_FORMAT = "dd/MM/yy HH:mm:ss"

    /**
     * Determines if a low battery location event is valid to be processed.
     *
     * This checks if a minimum time interval ([THREE_HOURS_MS]) has elapsed since the
     * last recorded event to prevent redundant triggers. If the event is valid, the
     * last event timestamp is updated to the current time.
     *
     */
    fun isValid(ctx: Context): Boolean {
        return try {
            val config = PreyConfig.getPreyConfig(ctx)
            val lastBatteryEventDate = config.locationLowBatteryDate
            val currentTime = System.currentTimeMillis()
            val sdf = SimpleDateFormat(DATE_FORMAT, Locale.getDefault())
            PreyLogger.d("EVENT lastEvent: $lastBatteryEventDate ${sdf.format(Date(lastBatteryEventDate))}")
            PreyLogger.d("EVENT now:       $currentTime ${sdf.format(Date(currentTime))}")
            PreyLogger.d("EVENT diff:      ${currentTime - lastBatteryEventDate}")
            if (lastBatteryEventDate == 0L || (currentTime - lastBatteryEventDate) > THREE_HOURS_MS) {
                config.locationLowBatteryDate = currentTime
                true
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

}